File: src/main/java/com/polarbookshop/dispatcherservice/DispatchingFunctions.java
package com.polarbookshop.dispatcherservice;


import org.slf4j.LoggerFactory;
import org.slf4j.Logger;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import reactor.core.publisher.Flux;

import java.util.function.Function;

@Configuration
class DispatchingFunctions {
    private static final Logger log = LoggerFactory.getLogger(DispatchingFunctions.class);

    @Bean
    public Function<OrderAcceptedMessage, Long> pack() {
        return orderAcceptMessage -> {
            log.info("The orer with id {} is packed.", orderAcceptMessage.orderId());
            return orderAcceptMessage.orderId();
        };
    }

    @Bean
    public Function<Flux<Long>, Flux<OrderDispatchedMessage>> label() {
        return orderFlux -> orderFlux.map(orderId -> {
            log.info("The order with id {} is labeled.", orderId);
            return new OrderDispatchedMessage(orderId);
        });
    }
}

File: src/main/kotlin/com/polarbookshop/dispatcherservice/DispatchingFunctions.kt
package com.polarbookshop.dispatcherservice

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.messaging.Message
import org.springframework.messaging.support.MessageBuilder
import reactor.core.publisher.Flux
import java.util.function.Function

@Configuration
class DispatchingFunctions {
    companion object {
        val log: Logger = LoggerFactory.getLogger(DispatchingFunctions::class.java)
    }

//    @Bean
//    fun pack() : (OrderAcceptedMessage) -> Long {
//        return {
//            log.info("The order with id ${it.orderId} is packed.")
//            it.orderId
//        }
//    }

//    @Bean
//    fun pack() : Function<OrderAcceptedMessage, Long> {
//        return Function {
//            log.info("The order with id ${it.orderId} is packed.")
//            it.orderId
//        }
//    }

//    @Bean
//    fun label() : (Flux<Long>) -> Flux<OrderDispatchedMessage> {
//        return {
//            orderFlux -> orderFlux.map {
//                orderId ->
//                    log.info("The order with id $orderId is labeled.")
//                    OrderDispatchedMessage(orderId)
//            }
//        }
//    }

//    @Bean
//    fun label() : Function<Flux<Long>, Flux<OrderDispatchedMessage>> {
//        return Function {
//            orderFlux -> orderFlux.map {
//                orderId ->
//                    log.info("The order with id $orderId is labeled.")
//                    OrderDispatchedMessage(orderId)
//            }
//        }
//    }

    @Bean
    fun pack(): (Message<OrderAcceptedMessage>) -> Message<Long> {
        return { message ->
            val orderId = message.payload.orderId
            log.info("The order with id $orderId is packed.")
            MessageBuilder.withPayload(orderId).build()
        }
    }

    @Bean
    fun label(): (Flux<Message<Long>>) -> Flux<Message<OrderDispatchedMessage>> {
        return { messageFlux ->
            messageFlux.map { message ->
                val orderId = message.payload
                log.info("The order with id $orderId is labeled.")
                MessageBuilder.withPayload(OrderDispatchedMessage(orderId)).build()
            }
        }
    }
}